cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/ProjectionTransform.java

```diff
@@ -0,0 +1,301 @@
+ */
+public class ProjectionTransform extends Transform<StructuredRecord, StructuredRecord> {
+  private static final String DROP = "drop";
+  private static final String RENAME = "rename";
+  private static final String CONVERT = "convert";
+  private static final Pattern fieldDelimiter = Pattern.compile("\\s*,\\s*");
+  private Set<String> fieldsToDrop = Sets.newHashSet();
+  private Map<String, String> fieldsToRename = Maps.newHashMap();
+  private Map<String, Schema.Type> fieldsToConvert = Maps.newHashMap();
+  // cache input schema hash to output schema so we don't have to build it each time
+  private Map<Schema, Schema> schemaCache = Maps.newHashMap();
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(getClass().getSimpleName());
+    configurer.setDescription("Projection transform that lets you drop, rename, and cast fields to a different type.");
+    configurer.addProperty(new Property(
+      DROP,
+      "Comma separated list of fields to drop. For example: 'field1,field2,field3'.",
+      false));
+    configurer.addProperty(new Property(
+      RENAME,
+      "List of fields to rename. This is a comma separated list of key-value pairs, where each " +
+        "pair is separated by a colon and specifies the input name and the output name. " +
+        "For example: 'datestr:date,timestamp:ts' specifies that the 'datestr' field should be renamed to " +
+        "'date' and the 'timestamp' field should be renamed to 'ts'.",
+      false));
+    configurer.addProperty(new Property(
+      CONVERT,
+      "List of fields to convert to a different type. This is a comma separated list of key-value pairs, where each " +
+        "pair is separated by a colon and specifies the field name and the desired type. " +
+        "For example: 'count:long,price:double' specifies that the 'count' field should be converted to a long " +
+        "and the 'price' field should be converted to a double. Only simple types are supported " +
+        "(boolean, int, long, float, double, bytes, string). Any simple type can be converted to bytes or a string. " +
+        "Otherwise, a type can only be converted to a larger type. For example, an int can be converted to a long, " +
+        "but a long cannot be converted to an int.",
+      false));
+  }
+
+  @Override
+  public void initialize(TransformContext context) {
+    Map<String, String> properties = context.getRuntimeArguments();
+
+    String dropStr = properties.get(DROP);
+    if (dropStr != null) {
+      for (String dropField : Splitter.on(fieldDelimiter).split(dropStr)) {
+        fieldsToDrop.add(dropField);
+      }
+    }
+
+    KeyValueListParser kvParser = new KeyValueListParser("\\s*,\\s*", ":");
+    String renameStr = properties.get(RENAME);
+    if (renameStr != null) {
+      for (KeyValue<String, String> keyVal : kvParser.parse(renameStr)) {
+        fieldsToRename.put(keyVal.getKey(), keyVal.getValue());
```

Comment:
Check if the value already exists before insert. If so that will be an initialization error. Renaming two source fields to a single target